homeassistant/components/abode/config_flow.py

```diff
@@ -26,48 +27,151 @@ def __init__(self):
+            if ex.errcode == MFA_CODE_REQUIRED[0]:
+                return await self.async_step_mfa()
+
             LOGGER.error("Unable to connect to Abode: %s", str(ex))
+
             if ex.errcode == HTTP_BAD_REQUEST:
-                return self._show_form({"base": "invalid_auth"})
-            return self._show_form({"base": "cannot_connect"})
-
-        return self.async_create_entry(
-            title=user_input[CONF_USERNAME],
-            data={
-                CONF_USERNAME: username,
-                CONF_PASSWORD: password,
-                CONF_POLLING: polling,
-            },
+                errors = {"base": "invalid_auth"}
+            else:
+                errors = {"base": "cannot_connect"}
+
+        if errors:
+            return self.async_show_form(
+                step_id=step_id, data_schema=vol.Schema(self.data_schema), errors=errors
+            )
+
+        return await self.async_step_finish(
+            {
+                CONF_USERNAME: self._username,
+                CONF_PASSWORD: self._password,
+                CONF_POLLING: self._polling,
+            }
         )
 
-    @callback
-    def _show_form(self, errors=None):
-        """Show the form to the user."""
-        return self.async_show_form(
-            step_id="user",
-            data_schema=vol.Schema(self.data_schema),
-            errors=errors if errors else {},
+    async def async_step_mfa(self, user_input=None):
+        """Handle multi-factor authentication (MFA)."""
+        errors = {}
+
+        if user_input is None:
+            return self.async_show_form(
+                step_id="mfa", data_schema=vol.Schema(self.mfa_data_schema)
+            )
+
+        try:
+            mfa_code = user_input[CONF_MFA]
+            abode = Abode(
+                auto_login=False,
+                get_devices=False,
+                get_automations=False,
+                cache_path=self._cache,
+            )
+
+            await self.hass.async_add_executor_job(
+                abode.login, self._username, self._password, mfa_code
+            )
+
+        except AbodeException as ex:
+            LOGGER.error("Invalid MFA code: %s", str(ex))
+            errors = {"base": "invalid_mfa_code"}
+
+        if errors:
+            return self.async_show_form(
+                step_id="mfa",
+                data_schema=vol.Schema(self.mfa_data_schema),
+                errors=errors,
+            )
+
+        return await self.async_step_finish(
+            {
+                CONF_USERNAME: self._username,
+                CONF_PASSWORD: self._password,
+                CONF_POLLING: self._polling,
+            }
         )
 
+    async def async_step_finish(self, user_input=None):
+        """Handle creating the config entry."""
+        existing_entry = await self.async_set_unique_id(self._username)
+
+        if existing_entry:
+            self.hass.config_entries.async_update_entry(existing_entry, data=user_input)
```

Comment:
Should we check that `user_input` isn't `None`?

Reply:
I thought the `vol.Required` took care of this as both the username and password field are required, so there's no way the `async_step_finish` would ever be called without some user input.

Reply:
Ok, then we should not default `user_input` to `None`.